pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/BrokerVersionFilter.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright 2016 Yahoo Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yahoo.pulsar.broker.loadbalance.impl;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.yahoo.pulsar.broker.BrokerData;
+import com.yahoo.pulsar.broker.loadbalance.BrokerFilter;
+import com.yahoo.pulsar.broker.BundleData;
+import com.yahoo.pulsar.broker.ServiceConfiguration;
+import com.yahoo.pulsar.broker.loadbalance.LoadData;
+
+public class BrokerVersionFilter implements BrokerFilter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GenericBrokerHostUsageImpl.class);
+
+    /**
+     * From the given set of available broker candidates, filter those using the load data.
+     *
+     * @param brokers
+     *            The currently available brokers that have not already been filtered. This set may be modified by
+     *            filter.
+     * @param bundleToAssign
+     *            The data for the bundle to assign.
+     * @param loadData
+     *            The load data from the leader broker.
+     * @param conf
+     *            The service configuration.
+     */
+    public void filter(Set<String> brokers, BundleData bundleToAssign, LoadData loadData, ServiceConfiguration conf) {
+        if ( conf.getActiveVersion() == null || conf.getActiveVersion().length() == 0 ) {
+            return;
+        }
+
+        final String activeVersion = conf.getActiveVersion();
+
+        Map<String, BrokerData> brokerData = loadData.getBrokerData();
+        List<String> brokersToRemove = new LinkedList<>();
+        for ( String broker : brokers ) {
+            BrokerData data = brokerData.get(broker);
+            if ( null == data ) {
+                LOG.warn("Cannot determine version of broker [{}]", broker);
+                continue;
+            }
+
+            String brokerVersion = data.getLocalData().getBrokerVersionString();
+            if ( null == brokerVersion || brokerVersion.length() == 0 ) {
+                LOG.warn("Broker [{}] load report does not contain a version string", broker);
+                continue;
+            }
+
+            LOG.debug("Broker [{}] is running version [{}]", broker, brokerVersion);
+            if ( !brokerVersion.equalsIgnoreCase(activeVersion)) {
```

Comment:
Unfortunately, [jsemver](https://github.com/zafarkhaja/jsemver) does not accept our current Pulsar version strings since it requires major, minor, and patch versions.

```
for ( String arg : args ) {
    try {
        Version version = Version.valueOf(arg);
        System.out.println("Valid: " + arg);
} catch ( Exception x ) {
    System.out.println("Invalid: " + arg + "(" + x.getMessage() + ")");
}
```

This will show that the third integer (the patch version) seems to be required.
```
Valid: 1.2.3
Invalid: 1.2(null)
Valid: 1.2.3-SNAPSHOT
Invalid: 1.2-SNAPSHOT(null)
```

I'll infer it to be "0" if it is missing and also submit a PR to [jsemver](https://github.com/zafarkhaja/jsemver) to hopefully get it fixed.